Owntracks test_encrypted_payload_libsodium fails

Make sure you are running the latest version of Home Assistant before reporting an issue.

You should only file an issue if you found a bug. Feature and enhancement requests should go in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum:

**Home Assistant release (`hass --version`):**
0.31.0.dev0

**Python release (`python3 --version`):**
Python 3.5.2

**Component/platform:**
device_tracker/owntracks

**Description of problem:**
TestDeviceTrackerOwnTracks.test_encrypted_payload_libsodium fails with KeyError during setup of device_tracker component.

**Expected:**
Test to not fail.

1. Make sure libsodium and libnacl is installed, otherwise the test will be skipped.
2. `py.test -v --cov-report term-missing --cov=homeassistant.components.device_tracker tests/components/device_tracker/`
3. See test fail with KeyError.

**Traceback (if applicable):**
```bash
________________________________________________________________________ TestDeviceTrackerOwnTracks.test_encrypted_payload_libsodium _________________________________________________________________________

self = <tests.components.device_tracker.test_owntracks.TestDeviceTrackerOwnTracks testMethod=test_encrypted_payload_libsodium>

    @unittest.skipUnless(libnacl, "libnacl/libsodium is not installed")
    def test_encrypted_payload_libsodium(self):
        """Test sending encrypted message payload."""
        self.assertTrue(device_tracker.setup(self.hass, {
            device_tracker.DOMAIN: {
                CONF_PLATFORM: 'owntracks',
>               CONF_SECRET: SECRET_KEY,
            }}))

tests/components/device_tracker/test_owntracks.py:698: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hass = <homeassistant.core.HomeAssistant object at 0x7fa98b3c5fd0>, config = {'device_tracker': {'platform': 'owntracks', 'secret': 's3cretkey'}}

    def setup(hass: HomeAssistantType, config: ConfigType):
        """Setup device tracker."""
        yaml_path = hass.config.path(YAML_DEVICES)
    
        try:
            conf = config.get(DOMAIN, [])
        except vol.Invalid as ex:
            log_exception(ex, DOMAIN, config)
            return False
        else:
>           conf = conf[0] if len(conf) > 0 else {}
E           KeyError: 0

homeassistant/components/device_tracker/__init__.py:120: KeyError
--------------------------------------------------------------------------------------------- Captured log setup ---------------------------------------------------------------------------------------------
selector_events.py          53 DEBUG    Using selector: EpollSelector
core.py                    457 INFO     Bus:Handling <Event service_registered[L]: service=publish, domain=mqtt>
common.py                  346 DEBUG    Configuration for zone, Validated: [], Original None
core.py                    457 INFO     Bus:Handling <Event component_loaded[L]: component=mqtt>
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=zone.home, new_state=<state zone.home=zoning; icon=mdi:home, radius=100, longitude=-117.22743, friendly_name=test home, hidden=True, latitude=32.87336 @ 2016-10-19T20:10:59.700381+00:00>, old_state=None>
core.py                    457 INFO     Bus:Handling <Event component_loaded[L]: component=zone>
loader.py                  141 INFO     Loaded device_tracker.owntracks from homeassistant.components.device_tracker.owntracks
bootstrap.py                43 DEBUG    Component mqtt already set up.
common.py                  346 DEBUG    Configuration for device_tracker, Validated: [{'waypoints': True, 'max_gps_accuracy': 200.0, 'consider_home': datetime.timedelta(0, 180), 'waypoint_whitelist': ['jon', 'greg'], 'track_new_devices': True, 'platform': 'owntracks'}], Original {'waypoints': True, 'max_gps_accuracy': 200, 'waypoint_whitelist': ['jon', 'greg'], 'platform': 'owntracks'}
bootstrap.py                43 DEBUG    Component mqtt already set up.
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, new_state=<state group.all_devices=unknown; entity_id=(), friendly_name=all devices, order=0, auto=True, hidden=True @ 2016-10-19T20:11:00.174994+00:00>, old_state=None>
core.py                    457 INFO     Bus:Handling <Event service_registered[L]: service=see, domain=device_tracker>
core.py                    457 INFO     Bus:Handling <Event component_loaded[L]: component=device_tracker>
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=zone.inner, new_state=<state zone.inner=zoning; name=zone, longitude=1.1, radius=10, latitude=2.1 @ 2016-10-19T20:11:00.218474+00:00>, old_state=None>
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=zone.inner_2, new_state=<state zone.inner_2=zoning; name=zone, longitude=1.1, radius=10, latitude=2.1 @ 2016-10-19T20:11:00.219566+00:00>, old_state=None>
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=zone.outer, new_state=<state zone.outer=zoning; name=zone, longitude=1.0, radius=100000, latitude=2.0 @ 2016-10-19T20:11:00.220778+00:00>, old_state=None>
core.py                    457 INFO     Bus:Handling <Event state_changed[L]: entity_id=device_tracker.greg_phone, new_state=<state device_tracker.greg_phone=None @ 2016-10-19T20:11:00.222009+00:00>, old_state=None>
```

**Additional info:**
My guess is that it has to do with the test trying to setup the `device_tracker` component a second time with a new config. The component is already set up in the test class setup method, and can't be set up more than once. We could try to move the test to the other test class that holds tests that need custom configs, where the class setup method doesn't set up the `device_tracker` component, but lets the test method do that.

CC @molobrakos

--- a/tests/components/device_tracker/test_owntracks.py
+++ b/tests/components/device_tracker/test_owntracks.py
@@ -2,18 +2,16 @@
 import json
 import os
 import unittest
+from collections import defaultdict
 from unittest.mock import patch
 
-from collections import defaultdict
+from tests.common import (assert_setup_component, fire_mqtt_message,
+                          get_test_home_assistant, mock_mqtt_component)
 
+import homeassistant.components.device_tracker.owntracks as owntracks
 from homeassistant.bootstrap import setup_component
 from homeassistant.components import device_tracker
-from homeassistant.const import (STATE_NOT_HOME, CONF_PLATFORM)
-import homeassistant.components.device_tracker.owntracks as owntracks
-
-from tests.common import (
-    assert_setup_component, get_test_home_assistant, mock_mqtt_component,
-    fire_mqtt_message)
+from homeassistant.const import CONF_PLATFORM, STATE_NOT_HOME
 
 USER = 'greg'
 DEVICE = 'phone'
@@ -684,23 +682,6 @@ class TestDeviceTrackerOwnTracks(BaseMQTT):
         new_wayp = self.hass.states.get(WAYPOINT_ENTITY_NAMES[0])
         self.assertTrue(wayp == new_wayp)
 
-    try:
-        import libnacl
-    except (ImportError, OSError):
-        libnacl = None
-
-    @unittest.skipUnless(libnacl, "libnacl/libsodium is not installed")
-    def test_encrypted_payload_libsodium(self):
-        """Test sending encrypted message payload."""
-        self.assertTrue(device_tracker.setup(self.hass, {
-            device_tracker.DOMAIN: {
-                CONF_PLATFORM: 'owntracks',
-                CONF_SECRET: SECRET_KEY,
-            }}))
-
-        self.send_message(LOCATION_TOPIC, ENCRYPTED_LOCATION_MESSAGE)
-        self.assert_location_latitude(2.0)
-
 
 class TestDeviceTrackerOwnTrackConfigs(BaseMQTT):
     """Test the OwnTrack sensor."""
@@ -803,3 +784,21 @@ class TestDeviceTrackerOwnTrackConfigs(BaseMQTT):
                     }}})
         self.send_message(LOCATION_TOPIC, MOCK_ENCRYPTED_LOCATION_MESSAGE)
         self.assert_location_latitude(None)
+
+    try:
+        import libnacl
+    except (ImportError, OSError):
+        libnacl = None
+
+    @unittest.skipUnless(libnacl, "libnacl/libsodium is not installed")
+    def test_encrypted_payload_libsodium(self):
+        """Test sending encrypted message payload."""
+        with assert_setup_component(1, device_tracker.DOMAIN):
+            assert setup_component(self.hass, device_tracker.DOMAIN, {
+                device_tracker.DOMAIN: {
+                    CONF_PLATFORM: 'owntracks',
+                    CONF_SECRET: SECRET_KEY,
+                    }})
+
+        self.send_message(LOCATION_TOPIC, ENCRYPTED_LOCATION_MESSAGE)
+        self.assert_location_latitude(2.0)

